### PR TITLE
Add Componenets to API

### PR DESCRIPTION
**Pull request by szok**
Added:
- DiskShareComponent
- DiskShareMountComponent
- FibreChannelComponent
- MemoryComponent
- OperatingSystemComponent
- ProcessorComponent
- SoftwareComponent

### CODE REVIEW

**Comment by ar4s** [comment]
:ok:

File: src/ralph/assets/api/serializers.py [hunk 1]
```diff
@@ -205,3 +206,76 @@ class Meta:
...
+        model = components.FibreChannelComponent
+
+
+class SoftwareComponentSerializer(ComponentSerializerMixin):
+
+    class Meta:
+        model = components.SoftwareComponent
+
+
+class SplunkUsageComponentSerializer(ComponentSerializerMixin):
```
Comment (mkurek) [inline src/ralph/assets/api/serializers.py]: we don't need this anymore

File: src/ralph/assets/api/serializers.py [hunk 2]
```diff
@@ -205,3 +206,76 @@ class Meta:
...
+        model = components.ProcessorComponent
+
+
+class MemoryComponentSerializer(ComponentSerializerMixin):
+
+    class Meta:
+        model = components.MemoryComponent
+
+
+class StorageComponentSerializer(ComponentSerializerMixin):
```
Comment (mkurek) [inline src/ralph/assets/api/serializers.py]: please remove it now - i don't know if we'll find usage for it

File: src/ralph/assets/models/components.py [hunk 3]
```diff
@@ -77,3 +78,210 @@ class GenericComponent(Component):
...
+        return '{} ({})'.format(self.label, self.physical_id)
+
+
+class SoftwareComponent(Component):
+    sn = models.CharField(
+        verbose_name=_('vendor SN'), max_length=255, unique=True, null=True,
+        blank=True, default=None,
+    )
+    label = models.CharField(verbose_name=_('name'), max_length=255)
+    # bash and widnows have a limit on the path length
```
Comment (mkurek) [inline src/ralph/assets/models/components.py]: windows

File: src/ralph/assets/models/components.py [hunk 4]
```diff
@@ -77,3 +78,210 @@ class GenericComponent(Component):
...
+class OperatingSystemComponent(Component):
+    label = models.CharField(verbose_name=_('name'), max_length=255)
+    memory = models.PositiveIntegerField(
+        verbose_name=_('memory'), help_text=_('in MiB'), null=True, blank=True,
+    )
+    storage = models.PositiveIntegerField(
+        verbose_name=_('storage'), help_text=_('in MiB'), null=True,
+        blank=True,
+    )
+    cores_count = models.PositiveIntegerField(
```
Comment (mkurek) [inline src/ralph/assets/models/components.py]: cores count is not necessary here

File: src/ralph/assets/api/views.py [hunk 5]
```diff
@@ -5,10 +5,60 @@
...
+class ComponentModelViewset(RalphAPIViewSet):
+    queryset = components.ComponentModel.objects.all()
+    serializer_class = serializers.ComponentModelSerializer
+
+
+class GenericComponentViewset(RalphAPIViewSet):
+    queryset = components.GenericComponent.objects.all()
+    serializer_class = serializers.GenericComponentSerializer
+    select_related = ['base_object', 'model']
+    prefetch_related = ['base_object__tags']
```
Comment (mkurek) [inline src/ralph/assets/api/views.py:58]: Shouldn't we fetch related tags in all viewsets for components? Why other viewsets are not inheriting from `GenericComponentViewset`?

File: src/ralph/assets/models/components.py [hunk 6]
```diff
@@ -77,3 +78,164 @@ class GenericComponent(Component):
...
+        verbose_name=_('share identifier'), null=True, blank=True,
+    )
+    label = models.CharField(
+        verbose_name=_('name'), max_length=255, blank=True, null=True,
+        default=None,
+    )
+    size = models.PositiveIntegerField(
+        verbose_name=_('size (MiB)'), null=True, blank=True,
+    )
+    snapshot_size = models.PositiveIntegerField(
```
Comment (mkurek) [inline src/ralph/assets/models/components.py:94]: Do we need this field?

File: src/ralph/assets/models/components.py [hunk 7]
```diff
@@ -77,3 +78,164 @@ class GenericComponent(Component):
...
+        verbose_name=_('size for snapshots (MiB)'), null=True, blank=True,
+    )
+    wwn = NullableCharField(
+        verbose_name=_('Volume serial'), max_length=33, unique=True,
+    )
+    full = models.BooleanField(default=True)
```
Comment (mkurek) [inline src/ralph/assets/models/components.py:100]: Again - what's the purpose of this field?

File: src/ralph/assets/models/components.py [hunk 8]
```diff
@@ -77,3 +78,164 @@ class GenericComponent(Component):
     class Meta:
         verbose_name = _('generic component')
         verbose_name_plural = _('generic components')
+
+
+class DiskShareComponent(Component):
+    share_id = models.PositiveIntegerField(
```
Comment (mkurek) [inline src/ralph/assets/models/components.py:84]: What's the purpose of this field?

File: src/ralph/assets/models/components.py [hunk 9]
```diff
@@ -77,3 +78,164 @@ class GenericComponent(Component):
...
+        verbose_name = _('disk share')
+        verbose_name_plural = _('disk shares')
+
+    def __str__(self):
+        return '%s (%s)' % (self.label, self.wwn)
+
+
+class DiskShareMountComponent(models.Model):
+    share = models.ForeignKey(DiskShareComponent, verbose_name=_('share'))
+    asset = models.ForeignKey(
```
Comment (mkurek) [inline src/ralph/assets/models/components.py:115]: Why not base object?

File: src/ralph/assets/models/components.py [hunk 10]
```diff
@@ -77,3 +78,164 @@ class GenericComponent(Component):
...
+        Asset, verbose_name=_('asset'), null=True, blank=True,
+        default=None, on_delete=models.SET_NULL
+    )
+    volume = models.CharField(
```
Comment (mkurek) [inline src/ralph/assets/models/components.py:119]: Maybe `volume_name`? Please verify how this field was used in Ralph 2

File: src/ralph/assets/models/components.py [hunk 11]
```diff
@@ -77,3 +78,164 @@ class GenericComponent(Component):
...
+class ProcessorComponent(Component):
+
+    label = models.CharField(verbose_name=_('name'), max_length=255)
+    speed = models.PositiveIntegerField(
+        verbose_name=_('speed (MHz)'), null=True, blank=True,
+    )
+    cores = models.PositiveIntegerField(
+        verbose_name=_('number of cores'), null=True, blank=True,
+    )
+    index = models.PositiveIntegerField(
```
Comment (mkurek) [inline src/ralph/assets/models/components.py:148]: If it's slot number why it's called index? Please verify it.

File: src/ralph/assets/models/components.py [hunk 12]
```diff
@@ -77,3 +78,164 @@ class GenericComponent(Component):
...
+        return '%s@%r' % (self.volume, self.asset)
+
+
+class ProcessorComponent(Component):
+
+    label = models.CharField(verbose_name=_('name'), max_length=255)
+    speed = models.PositiveIntegerField(
+        verbose_name=_('speed (MHz)'), null=True, blank=True,
+    )
+    cores = models.PositiveIntegerField(
```
Comment (mkurek) [inline src/ralph/assets/models/components.py:145]: Do there will be 3 places with cores count: asset model, component model and component. Please propose the solution where there is only one such place.

File: src/ralph/assets/models/components.py [hunk 13]
```diff
@@ -77,3 +78,164 @@ class GenericComponent(Component):
...
+    def __str__(self):
+        return '%r' % self.label
+
+
+class OperatingSystemComponent(Component):
+    label = models.CharField(verbose_name=_('name'), max_length=255)
+    memory = models.PositiveIntegerField(
+        verbose_name=_('memory'), help_text=_('in MiB'), null=True, blank=True,
+    )
+    storage = models.PositiveIntegerField(
```
Comment (mkurek) [inline src/ralph/assets/models/components.py:231]: I don't see any purpose of this field.

File: src/ralph/assets/models/components.py [hunk 14]
```diff
@@ -77,3 +78,164 @@ class GenericComponent(Component):
...
+    index = models.PositiveIntegerField(
+        verbose_name=_("slot number"), null=True, blank=True,
+    )
+
+    class Meta:
+        verbose_name = _('memory')
+        verbose_name_plural = _('memories')
+
+    def __str__(self):
+        return '#{}: {} ({})'.format(self.index, self.label, self.model)
```
Comment (ghost) [inline src/ralph/assets/models/components.py:187]:
```
return '#{}: {} ({})'.format(self.index or '-', self.label, self.model or '-')
```

?

File: src/ralph/assets/models/components.py [hunk 15]
```diff
@@ -77,3 +78,164 @@ class GenericComponent(Component):
...
+    index = models.PositiveIntegerField(
+        verbose_name=_('slot number'), null=True, blank=True,
+    )
+
+    class Meta:
+        verbose_name = _('processor')
+        verbose_name_plural = _('processors')
+
+    def __str__(self):
+        return '#{}: {} ({})'.format(self.index, self.label, self.model)
```
Comment (ghost) [inline src/ralph/assets/models/components.py:157]:
```
return '#{}: {} ({})'.format(self.index or '-', self.label, self.model or '-')
```

?

File: src/ralph/assets/models/components.py [hunk 16]
```diff
@@ -77,3 +78,164 @@ class GenericComponent(Component):
...
+    def get_size(self):
+        return self.size or self.share.get_total_size()
+
+    class Meta:
+        unique_together = ('share', 'asset')
+        verbose_name = _('disk share mount')
+        verbose_name_plural = _('disk share mounts')
+
+    def __str__(self):
+        return '%s@%r' % (self.volume, self.asset)
```
Comment (ghost) [inline src/ralph/assets/models/components.py:136]:
```
return '%s@%r' % (self.volume or '-', self.asset or '-')
```

?

File: src/ralph/assets/models/components.py [hunk 17]
```diff
@@ -77,3 +78,164 @@ class GenericComponent(Component):
...
+
+    def get_total_size(self):
+        return (self.size or 0) + (self.snapshot_size or 0)
+
+    class Meta:
+        verbose_name = _('disk share')
+        verbose_name_plural = _('disk shares')
+
+    def __str__(self):
+        return '%s (%s)' % (self.label, self.wwn)
```
Comment (ghost) [inline src/ralph/assets/models/components.py:110]:
```
return '%s (%s)' % (self.label or '-', self.wwn or '-')
```

?

**Pull request closed without merging**